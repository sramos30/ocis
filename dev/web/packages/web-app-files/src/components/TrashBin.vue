<template>
  <div>
    <app-bar :breadcrumbs="breadcrumbs" :has-bulk-actions="true" />
    <app-loading-spinner v-if="loadResourcesTask.isRunning" />
    <template v-else>
      <no-content-message
        v-if="isEmpty"
        id="files-trashbin-empty"
        class="files-empty"
        icon="delete-bin-5"
      >
        <template #message>
          <span v-if="noContentMessage">{{ noContentMessage }}</span>
          <span v-else v-translate>You have no deleted files</span>
        </template>
      </no-content-message>
      <resource-table
        v-else
        id="files-trashbin-table"
        v-model="selectedResources"
        class="files-table"
        :class="{ 'files-table-squashed': !sidebarClosed }"
        :fields-displayed="['name', 'ddate']"
        :are-paths-displayed="true"
        :are-thumbnails-displayed="false"
        :resources="paginatedResources"
        :are-resources-clickable="false"
        :header-position="fileListHeaderY"
        :sort-by="sortBy"
        :sort-dir="sortDir"
        @sort="handleSort"
      >
        <template #contextMenu="{ resource }">
          <context-actions v-if="isResourceInSelection(resource)" :items="selectedResources" />
        </template>
        <template #footer>
          <pagination :pages="paginationPages" :current-page="paginationPage" />
          <list-info
            v-if="paginatedResources.length > 0"
            class="oc-width-1-1 oc-my-s"
            :files="totalFilesCount.files"
            :folders="totalFilesCount.folders"
          />
        </template>
      </resource-table>
    </template>
  </div>
</template>

<script lang="ts">
import { mapGetters, mapMutations, mapState } from 'vuex'
import AppBar from './AppBar/AppBar.vue'
import ResourceTable from './FilesList/ResourceTable.vue'

import MixinFilesListFilter from '../mixins/filesListFilter'
import MixinResources from '../mixins/resources'
import MixinMountSideBar from '../mixins/sidebar/mountSideBar'

import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import ListInfo from './FilesList/ListInfo.vue'
import Pagination from './FilesList/Pagination.vue'
import ContextActions from './FilesList/ContextActions.vue'
import { useResourcesViewDefaults } from '../composables'
import { bus } from 'web-pkg/src/instance'
import { defineComponent } from '@vue/composition-api'
import { Resource } from '../helpers/resource'

export default defineComponent({
  name: 'TrashBin',

  components: {
    AppBar,
    ResourceTable,
    AppLoadingSpinner,
    NoContentMessage,
    ListInfo,
    Pagination,
    ContextActions
  },

  mixins: [MixinResources, MixinMountSideBar, MixinFilesListFilter],
  props: {
    breadcrumbs: { type: Array, default: () => [] },
    noContentMessage: {
      type: String,
      required: false,
      default: ''
    }
  },

  setup() {
    return {
      ...useResourcesViewDefaults<Resource, any, any[]>()
    }
  },

  computed: {
    ...mapState('Files', ['files']),
    ...mapGetters('Files', ['highlightedFile', 'selectedFiles', 'totalFilesCount']),
    ...mapState('Files/sidebar', { sidebarClosed: 'closed' }),

    isEmpty() {
      return this.paginatedResources.length < 1
    }
  },

  created() {
    this.loadResourcesTask.perform(this)

    const loadResourcesEventToken = bus.subscribe('app.files.list.load', (path) => {
      this.loadResourcesTask.perform(this, this.$route.params.item === path, path)
    })

    this.$on('beforeDestroy', () => {
      bus.unsubscribe('app.files.list.load', loadResourcesEventToken)
    })
  },

  methods: {
    ...mapMutations('Files', ['LOAD_FILES', 'CLEAR_CURRENT_FILES_LIST'])
  }
})
</script>
