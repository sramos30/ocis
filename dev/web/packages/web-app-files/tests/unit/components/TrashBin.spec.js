import { mount } from '@vue/test-utils'
import TrashBin from '@files/src/components/TrashBin.vue'
import { getStore, localVue, createFile } from '@files/tests/unit/components/components.setup.js'

const stubs = {
  'app-bar': true,
  'app-loading-spinner': true,
  'no-content-message': true,
  'resource-table': true,
  'context-actions': true,
  pagination: true,
  'list-info': true,
  'router-link': true
}

const appLoadingSpinnerStub = 'app-loading-spinner-stub'
const noContentStub = 'no-content-message-stub'
const filesTableStub = 'resource-table-stub'
const filesTableSelector = '#files-trashbin-table'
const contextActionsStub = 'context-actions-stub'
const listInfoStub = 'list-info-stub'
const paginationStub = 'pagination-stub'

describe('Trashbin component', () => {
  it('should show the app-loading-spinner component when the view is still loading', () => {
    const wrapper = getMountedWrapper({ loading: true })

    expect(wrapper.find(appLoadingSpinnerStub).exists()).toBeTruthy()
    expect(wrapper.find(noContentStub).exists()).toBeFalsy()
    expect(wrapper.find(filesTableStub).exists()).toBeFalsy()
  })

  describe('when the view is not loading anymore', () => {
    it('should show the no content message component if the paginated resources is empty', () => {
      const wrapper = getMountedWrapper()

      expect(wrapper.find(appLoadingSpinnerStub).exists()).toBeFalsy()
      expect(wrapper.find(filesTableStub).exists()).toBeFalsy()
      expect(wrapper.find(noContentStub).exists()).toBeTruthy()
    })

    describe('when length of the paginated resources is greater than zero', () => {
      const resourceList = [
        createFile({ id: '1234' }),
        createFile({ id: '5896' }),
        createFile({ id: '9856' })
      ]
      const wrapper = getMountedWrapper({
        paginatedResources: resourceList
      })

      it('should not show the no content message component', () => {
        expect(wrapper.find(noContentStub).exists()).toBeFalsy()
      })

      it('should load the resource table with correct props', () => {
        stubs['resource-table'] = false
        const wrapper = getMountedWrapper({
          paginatedResources: resourceList
        })
        const filesTable = wrapper.find(filesTableSelector)

        expect(filesTable.exists()).toBeTruthy()
        expect(filesTable).toMatchSnapshot()

        stubs['resource-table'] = true
      })

      describe('context menu', () => {
        let wrapper
        const selectedResources = [resourceList[0], resourceList[1]]
        const notSelectedResources = [resourceList[2]]
        beforeEach(() => {
          stubs['resource-table'] = false

          wrapper = getMountedWrapper({
            selectedFiles: selectedResources,
            paginatedResources: resourceList,
            paginationPage: 1,
            paginationPages: 2
          })
        })
        afterEach(() => {
          stubs['resource-table'] = true
        })
        it('should show the context actions for every selected resource', () => {
          selectedResources.forEach((selectedResource) => {
            const fileRow = wrapper.find(`[data-item-id="${selectedResource.id}"]`)
            const contextMenu = fileRow.find(contextActionsStub)
            expect(contextMenu.exists()).toBeTruthy()
            expect(contextMenu.props().items).toMatchObject(selectedResources)
          })
        })
        it('should not show the context actions for a resource that is not selected', () => {
          notSelectedResources.forEach((notSelectedResource) => {
            const fileRow = wrapper.find(`[data-item-id="${notSelectedResource.id}"]`)
            const contextMenu = fileRow.find(contextActionsStub)
            expect(contextMenu.exists()).toBeFalsy()
          })
        })
      })

      describe('pagination', () => {
        it('should be visible if the paginated pages is greater than zero', () => {
          stubs['resource-table'] = false
          const wrapper = getMountedWrapper({
            paginatedResources: resourceList,
            paginationPages: 2,
            paginationPage: 1
          })

          const pagination = wrapper.find(paginationStub)

          expect(pagination.exists()).toBeTruthy()
          expect(pagination.props()).toMatchObject({
            pages: 2,
            currentPage: 1
          })
          stubs['resource-table'] = true
        })
      })

      describe('list info', () => {
        it('should be present if the paginated resources list is not empty', () => {
          stubs['resource-table'] = false

          const wrapper = getMountedWrapper({
            paginatedResources: resourceList,
            paginationPage: 1,
            paginationPages: 2
          })

          const listInfo = wrapper.find(listInfoStub)
          expect(listInfo.exists()).toBeTruthy()
          expect(listInfo.props()).toMatchObject({
            files: 3,
            folders: 0
          })
          stubs['resource-table'] = true
        })
      })
    })
  })

  function mountOptions({ store, loading, paginatedResources, paginationPages, paginationPage }) {
    const $route = { params: { page: 1 } }
    const $router = {
      afterEach: jest.fn(),
      currentRoute: {
        name: 'some-route-name',
        query: {}
      },
      resolve: (r) => {
        return { href: r.name }
      }
    }
    return {
      localVue,
      store: store,
      stubs,
      mocks: {
        $route,
        $router
      },
      setup: () => ({
        loadResourcesTask: {
          isRunning: loading,
          perform: jest.fn()
        },
        paginatedResources: paginatedResources,
        paginationPages: paginationPages,
        paginationPage: paginationPage
      })
    }
  }

  function getMountedWrapper({
    selectedFiles = [],
    loading = false,
    paginatedResources = [],
    paginationPages = 12,
    paginationPage = 21
  } = {}) {
    const component = { ...TrashBin, created: jest.fn() }
    const store = createStore({
      totalFilesCount: { files: paginatedResources.length, folders: 0 },
      selectedFiles
    })
    return mount(
      component,
      mountOptions({
        store,
        loading,
        paginatedResources,
        paginationPages,
        paginationPage
      })
    )
  }

  function createStore({ totalFilesCount, highlightedFile, selectedFiles } = {}) {
    return getStore({
      highlightedFile,
      totalFilesCount,
      selectedFiles
    })
  }
})
