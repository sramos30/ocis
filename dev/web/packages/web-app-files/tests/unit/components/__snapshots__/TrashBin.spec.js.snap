// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trashbin component when the view is not loading anymore when length of the paginated resources is greater than zero should load the resource table with correct props 1`] = `
<table class="files-table oc-table oc-table-hover oc-table-sticky files-table-squashed" id="files-trashbin-table">
  <thead class="oc-thead">
    <tr tabindex="-1" class="oc-table-header-row">
      <th class="oc-th oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-text-nowrap oc-table-header-cell oc-table-header-cell-select oc-pl-m " style="top: 0px;"><span class="oc-table-thead-content"><div class="resource-table-select-all"><span><input id="resource-table-select-all" type="checkbox" name="checkbox" class="oc-checkbox oc-rounded oc-checkbox-l" value=""> <label for="resource-table-select-all" class="oc-invisible-sr oc-cursor-pointer">Select all resources</label></span></div></span>
        <!---->
      </th>
      <th class="oc-th oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-expand oc-text-nowrap oc-table-header-cell oc-table-header-cell-name" aria-sort="none" style="top: 0px;"><span class="oc-table-thead-content header-text">Name</span> <button aria-label="Sort by name" type="button" class="oc-button-sort oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-invisible"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span></button></th>
      <th class="oc-th oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-table-header-cell oc-table-header-cell-actions oc-pr-m" style="top: 0px;"><span class="oc-table-thead-content header-text">Actions</span>
        <!---->
      </th>
    </tr>
  </thead>
  <tbody>
    <tr tabindex="-1" data-item-id="file-id-1234" draggable="false" class="oc-tbody-tr oc-tbody-tr-file-id-1234">
      <td class="oc-td oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-table-data-cell oc-table-data-cell-select oc-pl-m "><span><input id="resource-table-select-file-id-1234" type="checkbox" name="checkbox" class="oc-checkbox oc-rounded oc-checkbox-l" value="[object Object]"> <label for="resource-table-select-file-id-1234" class="oc-invisible-sr oc-cursor-pointer">Select folder</label></span></td>
      <td class="oc-td oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-expand oc-text-truncate oc-table-data-cell oc-table-data-cell-name">
        <div class="resource-table-resource-wrapper">
          <div class="oc-resource oc-text-overflow"><span><span class="oc-icon oc-icon-l oc-icon-passive oc-resource-icon oc-resource-icon-file"><!----></span></span>
            <div class="oc-resource-details oc-text-overflow"><span class="oc-text-overflow"><!----> <span data-test-resource-path="/file-path/1234" data-test-resource-name="file-path/file-name-1234" data-test-resource-type="folder" class="oc-resource-name"><span class="oc-text-truncate"><span class="oc-resource-basename">file-name-1234</span></span>
              <!----></span></span>
              <div class="oc-resource-indicators">
                <router-link-stub to="[object Object]" class="parent-folder" style="cursor: pointer;"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span> <span class="text">file-path</span></router-link-stub>
              </div>
            </div>
          </div> <button type="button" class="resource-table-edit-name oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span></button>
        </div>
      </td>
      <td class="oc-td oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-table-data-cell oc-table-data-cell-actions oc-pr-m">
        <div class="resource-table-actions"> <button aria-label="Show context menu" type="button" class="resource-table-btn-action-dropdown oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" id="context-menu-trigger-file-id-1234"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></button>
          <div id="context-menu-drop-file-id-1234" class="oc-drop oc-box-shadow-medium oc-rounded"></div>
        </div>
      </td>
    </tr>
    <tr tabindex="-1" data-item-id="file-id-5896" draggable="false" class="oc-tbody-tr oc-tbody-tr-file-id-5896">
      <td class="oc-td oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-table-data-cell oc-table-data-cell-select oc-pl-m "><span><input id="resource-table-select-file-id-5896" type="checkbox" name="checkbox" class="oc-checkbox oc-rounded oc-checkbox-l" value="[object Object]"> <label for="resource-table-select-file-id-5896" class="oc-invisible-sr oc-cursor-pointer">Select folder</label></span></td>
      <td class="oc-td oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-expand oc-text-truncate oc-table-data-cell oc-table-data-cell-name">
        <div class="resource-table-resource-wrapper">
          <div class="oc-resource oc-text-overflow"><span><span class="oc-icon oc-icon-l oc-icon-passive oc-resource-icon oc-resource-icon-file"><!----></span></span>
            <div class="oc-resource-details oc-text-overflow"><span class="oc-text-overflow"><!----> <span data-test-resource-path="/file-path/5896" data-test-resource-name="file-path/file-name-5896" data-test-resource-type="folder" class="oc-resource-name"><span class="oc-text-truncate"><span class="oc-resource-basename">file-name-5896</span></span>
              <!----></span></span>
              <div class="oc-resource-indicators">
                <router-link-stub to="[object Object]" class="parent-folder" style="cursor: pointer;"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span> <span class="text">file-path</span></router-link-stub>
              </div>
            </div>
          </div> <button type="button" class="resource-table-edit-name oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span></button>
        </div>
      </td>
      <td class="oc-td oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-table-data-cell oc-table-data-cell-actions oc-pr-m">
        <div class="resource-table-actions"> <button aria-label="Show context menu" type="button" class="resource-table-btn-action-dropdown oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" id="context-menu-trigger-file-id-5896"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></button>
          <div id="context-menu-drop-file-id-5896" class="oc-drop oc-box-shadow-medium oc-rounded"></div>
        </div>
      </td>
    </tr>
    <tr tabindex="-1" data-item-id="file-id-9856" draggable="false" class="oc-tbody-tr oc-tbody-tr-file-id-9856">
      <td class="oc-td oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-table-data-cell oc-table-data-cell-select oc-pl-m "><span><input id="resource-table-select-file-id-9856" type="checkbox" name="checkbox" class="oc-checkbox oc-rounded oc-checkbox-l" value="[object Object]"> <label for="resource-table-select-file-id-9856" class="oc-invisible-sr oc-cursor-pointer">Select folder</label></span></td>
      <td class="oc-td oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-expand oc-text-truncate oc-table-data-cell oc-table-data-cell-name">
        <div class="resource-table-resource-wrapper">
          <div class="oc-resource oc-text-overflow"><span><span class="oc-icon oc-icon-l oc-icon-passive oc-resource-icon oc-resource-icon-file"><!----></span></span>
            <div class="oc-resource-details oc-text-overflow"><span class="oc-text-overflow"><!----> <span data-test-resource-path="/file-path/9856" data-test-resource-name="file-path/file-name-9856" data-test-resource-type="folder" class="oc-resource-name"><span class="oc-text-truncate"><span class="oc-resource-basename">file-name-9856</span></span>
              <!----></span></span>
              <div class="oc-resource-indicators">
                <router-link-stub to="[object Object]" class="parent-folder" style="cursor: pointer;"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span> <span class="text">file-path</span></router-link-stub>
              </div>
            </div>
          </div> <button type="button" class="resource-table-edit-name oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span></button>
        </div>
      </td>
      <td class="oc-td oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-table-data-cell oc-table-data-cell-actions oc-pr-m">
        <div class="resource-table-actions"> <button aria-label="Show context menu" type="button" class="resource-table-btn-action-dropdown oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" id="context-menu-trigger-file-id-9856"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></button>
          <div id="context-menu-drop-file-id-9856" class="oc-drop oc-box-shadow-medium oc-rounded"></div>
        </div>
      </td>
    </tr>
  </tbody>
  <tfoot class="oc-table-footer">
    <tr class="oc-table-footer-row">
      <td colspan="3" class="oc-table-footer-cell">
        <pagination-stub pages="12" currentpage="21"></pagination-stub>
        <list-info-stub files="3" folders="0" class="oc-width-1-1 oc-my-s"></list-info-stub>
      </td>
    </tr>
  </tfoot>
</table>
`;
